### PR TITLE
Example Executable

### PR DESCRIPTION
**Pull request by mattmassicotte**
This is a concept of an example-hosting executable. It includes some static variable usage to help get a feel for how something like this could be structured.

I really wanted to construct a `Package.swift` that could build the executable with three possible configurations:

- Swift 5 compiler (5 mode)
- Swift 6 compiler (5 mode)
- Swift 6 compiler (6 mode)

This turned out to be pretty tricky to pull off, but I have it working. SPM will not let you use the same path for multiple targets, but I also was able to avoid duplicating the code by using symlinks.

I'd like to get feedback on pretty much every part of this.

- How do you like the structure?
- Is maintaining compatibility with a Swift 5.10 toolchain a goal? It makes the `Package.swift` more complex.
- Does the use of conditional compilation in the sources seem like it makes sense?
- How hard should I be trying to ensure that data races actually occur in unsafe code?
- Is the level of discussion in the comments sufficient?

### CODE REVIEW

**Comment by mattmassicotte** [comment]
I'm encorporating some more changes here for boundary checking. And this is making me realize that how I have structured the targets doesn't quite line up with how to best illustrate the problems...

File: Package.swift [hunk 1]
```diff
@@ -2,43 +2,68 @@
...
+#if compiler(>=6.0)
+// this can be further improved once SE-0435 is available
+let swift5Mode = SwiftSetting.unsafeFlags(["-swift-version", "5"])
+let swift6Mode = SwiftSetting.unsafeFlags(["-swift-version", "6"])
+let toolsSwiftMode = swift6Mode
+let baseSettings: [SwiftSetting] = [
+]
+#else
+let swift5Mode = SwiftSetting.unsafeFlags(["-swift-version", "5"])
+let swift6Mode = SwiftSetting.unsafeFlags(["-swift-version", "6"])
```
Comment (hborla) [inline Package.swift:14]: Do these variables need to be included in each condition, or can they be lifted above? They appear to be the same.

Comment (mattmassicotte) [reply Package.swift:14]: Today they *must* be the same because there doesn't yet appear to be a Swift 6 snapshot that allows me to use `SwiftSetting. swiftLanguageVersion(.v6)`. I thought it could be nice to express the setting more semantically in that case.

I just checked the June 4 build and it still isn't working. But I also cannot tell *why* isn't working because of https://github.com/apple/swift/issues/74081

Comment (hborla) [reply Package.swift:14]: I'm just wondering why it isn't written like this

```swift
let swift5Mode = SwiftSetting.unsafeFlags(["-swift-version", "5"])
let swift6Mode = SwiftSetting.unsafeFlags(["-swift-version", "6"])

#if compiler(>=6.0)
// this can be further improved once SE-0435 is available
let toolsSwiftMode = swift6Mode
let baseSettings: [SwiftSetting] = [
]
#else
let toolsSwiftMode = swift5Mode
let baseSettings: [SwiftSetting] = [
	.enableExperimentalFeature("StrictConcurrency")
]
#endif
```

Comment (mattmassicotte) [reply Package.swift:14]: I was kinda hoping that eventually it could look like this to better highlight SE-0435. But, maybe this is silly. It does make it less concise...

```swift
#if compiler(>=6.0)
let swift5Mode = SwiftSetting.swiftLanguageVersion(.v5)
let swift6Mode = SwiftSetting.swiftLanguageVersion(.v5)
let toolsSwiftMode = swift6Mode
let baseSettings: [SwiftSetting] = [
]
#else
let swift5Mode = SwiftSetting.unsafeFlags(["-swift-version", "5"])
let swift6Mode = SwiftSetting.unsafeFlags(["-swift-version", "6"])
let toolsSwiftMode = swift5Mode
let baseSettings: [SwiftSetting] = [
	.enableExperimentalFeature("StrictConcurrency")
]
#endif
```

Comment (hborla) [reply Package.swift:14]: Ahh I understand!

File: Sources/Library/Library.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import Foundation
+
+/// An example of a struct with only `Sendable` properties.
+///
+/// This type is implicitly-`Sendable` within its defining module.
```
Comment (hborla) [inline Sources/Library/Library.swift:5]: Ah, this isn't true. There's no such thing as internal-only conformances, so this struct isn't implicitly `Sendable` at all. That said, inferring `Sendable` only within the module would be an interesting future direction!

Comment (mattmassicotte) [reply Sources/Library/Library.swift:5]: Ohhh my mistake! I did not realize that making them public removes the implict conformance. Shoot, I think I have some written content to update too now...

File: Sources/Examples/Boundaries.swift [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+// MARK: Computed Value
+
+/// A Sendable function is used to compute the value in a different isolation domain.
+func computedValue_updateStyle(using backgroundColorProvider: @Sendable () -> ColorComponents) async {
+	// pass backgroundColorProvider into the MainActor here
+	await MainActor.run {
+		// invoke it in this domain to actually create the needed value
+		let components = backgroundColorProvider()
+		applyBackground(components)
```
[the remark this answers is not in the document]
Comment (mattmassicotte) [reply Sources/Examples/Boundaries.swift:53]: What do you think now? I added a conditional to show both versions side by side. It really is amazing this is possible and I think it's a great thing to highlight.

File: Package.swift (hunk 1 above)
Comment (mattmassicotte) [reply Package.swift:14]: No I'm wrong about this. We cannot ever simultaneously use a 6.0 tools feature and also declare compatibility with 5.10.

Comment (mattmassicotte) [reply Package.swift]: I thought more on this. I think having Swift 5 targets is very useful, but I just don't see what we get by maintaining compatibility with Swift 5.10 tools. Plus it makes the package so much simpler.

File: Package.swift [hunk 4]
```diff
@@ -1,44 +1,44 @@
...
-                "FullyMigratedModule",
-                "MigrationInProgressModule",
-            ]
-        ),
-    ],
-    targets: [
-        .target(
-            name: "FullyMigratedModule"
-        ),
+	name: "MigrationGuide",
```
Comment (ktoso) [inline Package.swift:6]: nitpick about the tabs, especially since mixing tabs and spaces in same file.

Stick to spaces? No strong feels about their number but let's not mix spaces and tabs :)

File: Sources/Examples/Boundaries.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (ktoso) [reply Sources/Examples/Boundaries.swift:53]: I think the side by side is okey -- it'd need some text in the guide somewhere about how not resorting to MainActor.run is generally better, I see people use it too much sometimes but it's ok to show like this!

File: Sources/Examples/DispatchQueue+PendingWork.swift [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import Dispatch
+
+extension DispatchQueue {
+	/// Returns once any pending work has been completed.
+	func pendingWorkComplete() async {
+		await withUnsafeContinuation { continuation in
```
Comment (ktoso) [inline Sources/Examples/DispatchQueue+PendingWork.swift:6]: Please always show `withCheckedContinuation` 

```suggestion
		await withCheckedContinuation { continuation in
```

unless we're specifically discussing a reason to use the unsafe one

File: Sources/Examples/Globals.swift [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+#if swift(<6.0)
+/// An unsafe global variable.
+///
+/// See swift-6-concurrency-migration-guide/commonproblems/#Sendable-Types
+var supportedStyleCount = 42
+#endif
+
+/// Version of `supportedStyleCount` that uses global-actor isolation.
+@MainActor
+var globallyIsolated_supportedStyleCount = 42
```
Comment (ktoso) [inline Sources/Examples/Globals.swift:12]: nitpick but "globally isolated" is a term we don't have, perhaps just "`globalActorIsolated_...`"?

File: Sources/Library/Library.swift [hunk 7]
```diff
@@ -0,0 +1,41 @@
+import Foundation
+
+/// An example of a struct with only `Sendable` properties.
+///
+/// This type would be implicitly-`Sendable` within its defining module, but
+/// because it is public the `Sendable` conformance must be explicit.
+public struct ColorComponents {
```
Comment (ktoso) [inline Sources/Library/Library.swift:7]: The comment here is a bit confusing, because we're saying "it would be ... but ... so ... it must be explicitly Sendable" -- but then it's not actually Sendable. So we're explaining it "must"  but then we don't follow that must.

It makes me think "is this missing a Sendable, or not making it sendable on purpose..."

Maybe reword that: 

> this type is NOT Sendable because it is public. If we want a public type to be Sendable, we must annotate it explicitly.

?

File: Sources/Examples/Globals.swift [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+	manualSerialQueue.async {
+		queueProtected_supportedStyleCount += 1
+	}
+
+	manualSerialQueue.async {
+		print("  - accessing queueProtected_supportedStyleCount: ", queueProtected_supportedStyleCount)
+	}
+
+	await manualSerialQueue.pendingWorkComplete()
+}
```
Comment (ktoso) [inline Sources/Examples/Globals.swift:59]: (general nitpick that some files use spaces, some tabs -- let's pick one?)

File: Sources/Examples/Boundaries.swift [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+	}
+#else
+	// The Swift 6 compiler can automatically determine this value is
+	// being transferred in a safe way
+	let components = backgroundColorProvider()
+	await applyBackground(components)
+#endif
+}
+
+// MARK: Global Isolation
```
Comment (ktoso) [inline Sources/Examples/Boundaries.swift:63]:
```suggestion
// MARK: Global Actor Isolation
```

I'm concerned about saying "global isolation" isolation is always "to some actor", let's stick to that

**Review by ktoso — APPROVED** [review]
Pretty good! minor nitpicks, lgtm

File: Package.swift (hunk 4 above)
Comment (hborla) [reply Package.swift:6]: I too prefer spaces :)

Comment (mattmassicotte) [reply Package.swift:6]: I figured this would come up so I just quickly wanted to elaborate on why.

I used to be strongly pro-spaces. But I was one day informed that tabs provides a major quality-of-life impovement for low-vision readers that use large font sizes. So I began making a consious effort to switch.

I screw this up all the time, however, because I always reformat to spaces for DocC content, as it does not handle semantic indentation well. While I use tabs for all my own open source stuff, I regularly get contributions that use spaces and I do not nitpick over whitespace. I assume that'll be a problem here too. FWIW, TSPL uses 4-space indentation.

I've currently reformatted the code to at least be consistent. But, I am very happy to change, since there are pretty legimate technical hurdles here, on top of preferences.

File: Sources/Examples/DispatchQueue+PendingWork.swift (hunk 5 above)
Comment (mattmassicotte) [reply Sources/Examples/DispatchQueue+PendingWork.swift:6]: I have no idea why I did this. I don't think I have ever used the usafe variants before...

Comment (mattmassicotte) [reply Sources/Examples/DispatchQueue+PendingWork.swift:6]: Umm, so funny story...

https://github.com/apple/swift/issues/74206

File: Sources/Examples/Globals.swift (hunk 6 above)
Comment (mattmassicotte) [reply Sources/Examples/Globals.swift:12]: I think consistent, correct terminology is really important!

File: Sources/Examples/Boundaries.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattmassicotte) [reply Sources/Examples/Boundaries.swift:53]: Great!

We also have strengthened the wording around the disadvantages of dynamic isolation in the guide, though this particular construct has not been called out. Perhaps we should...

File: Sources/Examples/DispatchQueue+PendingWork.swift (hunk 5 above)
Comment (mattmassicotte) [reply Sources/Examples/DispatchQueue+PendingWork.swift:7]: I added an in-code TODO, because I cannot currently fix this.
